File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.java
import java.util.*;
import java.io.*;


public class Main{

    static int[][] graph;
    static boolean[] visited;

    static int N,M;


    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        int counter = 0;
        graph = new int[N+1][N+1];
        visited = new boolean[N+1];
        for(int i=0; i<M; i++){
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            graph[a][b] = graph[b][a] = 1;
        }
        for(int i=1; i<=N; i++){
            if(!visited[i]){
                dfs(i);
                counter++;
            }
        }
        System.out.println(counter);
    }

    static void dfs(int node){
        visited[node] = true;
        for(int i=1; i<=N; i++){
            if(graph[node][i] == 1 && !visited[i]){
                dfs(i);
            }
        }
    }
}

File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.js
const readline = require("readline");

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let inputLine = 0;
let arr;
let visited;
let N;
let M;
let answer = 0;
rl.on("line", (line) => {
  if (inputLine == 0) {
    [N, M] = line.split(" ").map((el) => parseInt(el));
    arr = Array.from({ length: N + 1 }, () => []);
    visited = Array.from({ length: N + 1 }, () => false);
  }
  if (inputLine > 0 && inputLine <= M) {
    let [a, b] = line.split(" ").map((el) => parseInt(el));
    arr[a].push(b);
    arr[b].push(a);
  }
  if (inputLine == M) {
    solution();
    console.log(answer);
    process.exit();
  }
  inputLine++;
});

function solution() {
  for (let i = 1; i < visited.length; i++) {
    if (!visited[i]) {
      dfs(i);
      answer++;
    }
  }
}

function dfs(node) {
  visited[node] = true;
  for (const next of arr[node]) {
    if (!visited[next]) {
      dfs(next);
    }
  }
}
